modules/nlu/src/backend/application/training-queue.ts

```diff
@@ -0,0 +1,287 @@
+import * as sdk from 'botpress/sdk'
+import * as NLU from 'common/nlu/engine'
+import _ from 'lodash'
+
+import moment from 'moment'
+import ms from 'ms'
+import nanoid from 'nanoid'
+import { ITrainingRepository, ITransactionContext } from './training-repo'
+import { TrainingId, TrainerService, TrainingListener, TrainingState, TrainingSession, I } from './typings'
+
+export interface TrainingQueueOptions {
+  maxTraining: number
+}
+
+const DEFAULT_OPTIONS: TrainingQueueOptions = {
+  maxTraining: 2
+}
+
+const DEFAULT_STATE = { status: <sdk.NLU.TrainingStatus>'idle', progress: 0 }
+
+export type ITrainingQueue = I<TrainingQueue>
+
+const debug = DEBUG('nlu').sub('lifecycle')
+const TRAINING_LIFE_TIME = ms('5m')
+
+const STATES_REQUIRING_NODE_AFFINITY: sdk.NLU.TrainingStatus[] = ['training', 'canceled', 'errored']
+
+export class TrainingQueue {
+  private _options: TrainingQueueOptions
+  private _paused: boolean = true
+
+  private _workerId = nanoid() // TODO: find a way to make this stay between server restart
+
+  constructor(
+    private _trainingRepo: ITrainingRepository,
+    private _errors: typeof NLU.errors,
+    private _logger: sdk.Logger,
+    private _trainerService: TrainerService,
+    private _onChange: TrainingListener,
+    options: Partial<TrainingQueueOptions> = {}
+  ) {
+    this._options = { ...DEFAULT_OPTIONS, ...options }
+  }
+
+  public async initialize() {
```

Comment:
IMO it's not TraininQueue's business to be aware that trainingRepo has to be initialized .. and should come initialized

Reply:
kind of arbitrary, but sure we can make all initialization when bootstrapping